### PR TITLE
build deps should respect skip-recommended

### PR DESCRIPTION
**Pull request by ilovezfs**
For example, `brew deps libass --skip-recommended` shouldn't print
harfbuzz because, even though libass builds with harfbuzz when harfbuzz
is not skipped, we asked to skip recommended, of which harfbuzz is one.

### CODE REVIEW

File: Library/Homebrew/cmd/deps.rb [hunk 1]
```diff
@@ -72,10 +72,10 @@ def deps_for_formula(f, recursive = false)
 
     if recursive
       deps = f.recursive_dependencies do |dependent, dep|
-        Dependency.prune if ignores.any? { |ignore| dep.send(ignore) } && !includes.any? { |include| dep.send(include) } && !dependent.build.with?(dep)
+        Dependency.prune if ignores.any? { |ignore| dep.send(ignore) } && !includes.any? { |include| dep.send(include) } && (!dependent.build.with?(dep) || ignores.any? { |ignore| dep.send(ignore) })
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/deps.rb]: I think it would be good to break this into multiple lines/statements. It was unwieldy before and it's a bit crazy now.

Comment (DomT4) [inline Library/Homebrew/cmd/deps.rb]: 👍

Comment (xu-cheng) [inline Library/Homebrew/cmd/deps.rb]: :+1: for mike's advice on breaking to multiple lines. (Current format can be a little hard to follow)

Also you probably want to update `brew used` as well

Comment (ilovezfs) [inline Library/Homebrew/cmd/deps.rb]: Actually I think we can just remove the `&& !dependent.build.with?(dep)`

```
Dependency.prune if ignores.any? { |ignore| dep.send(ignore) } && !includes.any? { |include| dep.send(include) } 
```

**Comment by xu-cheng** [comment]
Actually, now after some thought, I think current existing behavior is correct. `brew uses/deps` sort of behaves like `brew reinstall/upgrade`. So the current installation should be taken into the account. For example, if I `brew install foo --with-optional-bar`. `brew deps foo` should still return bar, even when the default is to filter optional deps.

**Comment by ilovezfs** [comment]
That would mean `--skip-recommended` basically does nothing unless it's `:run`.

**Comment by ilovezfs** [comment]
Perhaps we need to revert to the old behavior that had it in the reverse, --skip-optional and --include-recommended because --skip-recommended isn't useful right now, if you're saying the behavior is actually correct.

**Comment by xu-cheng** [comment]
To be more clearly, local installation should be treated with higher priority than `--skip/include` flags. The latter is used more often to determine deps before you install it.

**Comment by ilovezfs** [comment]
As far as I can tell the existing logic has nothing to do with local installation.

**Comment by ilovezfs** [comment]
Look at output of `brew deps libass --skip-recommended` and look at the `libass` formula, and tell me that makes sense!

**Comment by xu-cheng** [comment]
> because --skip-recommended isn't useful right now'

It can be useful if you want to do: `brew install $(brew --skip-recommended foo); brew install foo --without-recommended-deps`

> As far as I can tell the existing logic has nothing to do with local installation.

The `dependant.build` will check local installation.

**Comment by ilovezfs** [comment]
I'm looking locally at the command output in the absence of having these formulae installed. I thought `brew deps` and `brew uses` were currently supposed to be stateless commands having nothing to do with the user's local Cellar and that the output should match between systems.

**Comment by mikemcquaid** [comment]
> I thought brew deps and brew uses were currently supposed to be stateless commands having nothing to do with the user's local Cellar and that the output should match between systems.

That may not be the case but I agree it should be.

**Comment by xu-cheng** [comment]
Sorry, I made a mistake here. You are right that the existing logic won't check for local installation. What will check is the flag declared in formula. e.g. `depends_on: foo => [with-bar, without-baz]`. In this case, the build flags(declared by formula authors) should be still considered with higher priority than user's input flag.

**Comment by ilovezfs** [comment]
```

  depends_on "pkg-config" => :build
  depends_on "yasm" => :build

  depends_on "freetype"
  depends_on "fribidi"
  depends_on "harfbuzz" => :recommended
  depends_on "fontconfig" => :optional
```

```
iMac-TMP:~ joe$ brew deps --skip-recommended libass 
cairo
fontconfig
freetype
fribidi
gettext
glib
gobject-introspection
harfbuzz
icu4c
libffi
libpng
pixman
pkg-config
```

**Comment by xu-cheng** [comment]
I think the problem is because we don't take into account whether a filter(include or skip) is enforced by users or enforced by Homebrew's default behavior.

What should happen should be this:
```
skip if user explicitly passes --skip- flag
keep if user explicitly passes --include- flag
keep if the defendant should be built with it (declared by formula)
skip if it's optional (default behavior)
keep if it's recommended (default behavior)
```

**Comment by xu-cheng** [comment]
En, this question is not very straight forward. I probably need to take more time to figure this out the correct behaviors.

**Comment by ilovezfs** [comment]
The output with various combinations of the flags made sense to me with the "build veto" removed but take your time thinking about it.

**Comment by xu-cheng** [comment]
After some thought, I think this may be what we want.

```ruby
f.recursive_dependencies do |dependent, dep| do
  if dep.build?
    Dependency.prune unless ARGV.include? "--include-build"
  elsif dep.recommended?
    Dependency.prune if ARGV.include?("--skip-recommended") || dependent.build.without?(dep)
  elsif dep.optional?
    Dependency.prune if !ARGV.include?("--include-optional") && !dependent.build.with?(dep)
  end
end

**Comment by ilovezfs** [comment]
@xu-cheng what does it do in case of `:run` / `dep.run?`?

**Comment by xu-cheng** [comment]
`dep.run?` has nothing to do with this. As it's used only as a tag for `brew audit` or bottle system.

**Comment by mikemcquaid** [comment]
@xu-cheng That looks right to me.

**Comment by mikemcquaid** [comment]
@ilovezfs Any news here?

**Comment by ilovezfs** [comment]
@mikemcquaid @xu-cheng OK, I refreshed the PR with @xu-cheng's proposal. There are still some weird results, though. `--include-optional` seems to get extended recursively. So `brew deps lnav` gives you `pcre` and `readline`, with no mention of curl because
```
depends_on "curl" => ["with-libssh2", :optional]
```
OK, that makes sense. So let's say we want to see that `curl` show up, and make sure it's using `with-libssh2`.
```
brew deps --include-optional lnav
```
which now drags in all of curl's optional dependencies as well, yielding;
```
$ brew deps lnav --include-optional
atk
berkeley-db4
boost
c-ares
cairo
curl
doxygen
fontconfig
freetype
gdbm
gdk-pixbuf
gettext
glib
gobject-introspection
graphite2
graphviz
gsettings-desktop-schemas
gtk+3
gts
harfbuzz
hicolor-icon-theme
homebrew/dupes/tcl-tk
icu4c
jansson
jasper
jpeg
libcroco
libepoxy
libev
libevent
libffi
libgsf
libidn
libmetalink
libpng
libressl
librsvg
libssh2
libtiff
libxml2
netpbm
nghttp2
openssl
pango
pcre
pixman
pkg-config
python
readline
rtmpdump
spdylay
sqlite
```
This isn't a regression caused by this PR, but I do think we need to reconsider not showing first level optional by default since the output of `--include-optional` is now pretty over the top, and passing `-1` will hide the `libssh2` altogether.

**Comment by UniqMartin** [comment]
> There are still some weird results, though. `--include-optional` seems to get extended recursively.

How are you proposing to resolve this issue? I'd find it weird and hard to understand if `--include-optional` only applied to the immediate dependencies. And if we added a new option like `--include-optional-for=<formula1>,<formula2>,…`, our already hard to grasp `brew deps` with all its options would become even more complicated. And we'd need to do something similar for `--include-build` and `--skip-recommended`.

I'd say the list `brew deps` returns is more useful for consumption by scripts and only really simple queries by humans. For more complicated cases like the one you sketched, a human would need to look at the dependency tree (or graph), thus I'd suggest to improve `deps --tree` for this. It's functionality is currently very limited, but it shouldn't be too difficult to make it output a tree where build, optional, recommended, and hard dependencies are all printed together, but are still distinguishable. This form of visualization is also much better suited for adding something like `--max-levels=<n>` where the tree would be truncated at depth `<n>`, but leaves that haven't been fully expanded could be annotated as such.

**Comment by ilovezfs** [comment]
Something that does this (but with `deps` command, I'd think):
```
iMac-TMP:local joe$ brew install --with-curl lnav
readline
pcre
pkg-config
openssl
libssh2
curl
Error: OK
```
which is generated using
```
diff --git a/Library/Homebrew/formula_installer.rb b/Library/Homebrew/formula_installer.rb
index 4b8dcf2..f93a1e6 100644
--- a/Library/Homebrew/formula_installer.rb
+++ b/Library/Homebrew/formula_installer.rb
@@ -268,6 +268,8 @@ class FormulaInstaller
     check_requirements(req_map)
     deps = expand_dependencies(req_deps + formula.deps)
 
+    puts deps
+raise "OK"
     deps
   end
 
@@ -354,8 +356,8 @@ class FormulaInstaller
         Dependency.prune
       elsif dep.build? && install_bottle_for?(dependent, build)
         Dependency.prune
-      elsif dep.satisfied?(options)
-        Dependency.skip
+#      elsif dep.satisfied?(options)
+#        Dependency.skip
       end
     end
 

```

**Comment by ilovezfs** [comment]
Bump.

**Comment by UniqMartin** [comment]
I think the PR as-is is fine and good to go.

The discussion about how to make the output of `--include-optional` more useful with extensive dependency trees like in the `lnav` case probably needs to continue somewhere else. (I still believe that `brew deps --tree` can be more useful in this case, if it is extended a bit to respect the `--include-*` and `--skip-*` options and to annotate its output with “optional” etc.)

**Comment by ilovezfs** [comment]
@UniqMartin on a related note to the formula-specific-options thing, it occurred to me the other day that it's related to the same usage attempted here for nearly the same purposes https://github.com/Homebrew/legacy-homebrew/issues/48986#issuecomment-215114587

**Comment by ilovezfs** [comment]
Perhaps despite the redundancy, the formula specific options thing might be more trivially added to `brew install` as demonstrated above. Effectively, that's a `brew install --dry-run` for the sake of spitting out the dependencies that will actually be used, modulo whether to include or to exclude what's already satisfied on a given system.

**Comment by UniqMartin** [comment]
A `brew install --dry-run` (or some other command that has the same effect) makes sense to me. In the end, all of this just shows that our dependency handling and dependency resolution isn't as flexible and as hackable as it could be. I'm not sure at which point it is preferable to reimplement it from scratch (and with a proper design phase preceding it) instead of just tweaking various commands that try to use it. But this is certainly not an easy task to tackle …

**Comment by ilovezfs** [comment]
I think the `--dry-run` thing for `install` helps to distinguish between the questions that each command is trying to answer. `brew deps` is context free, describing the tree of dependencies and requirements from a universal perspective. But I think often the user is trying to answer a very context specific question: if I install foo with a given option set what specific set of deps/reqs is both necessary and sufficient to accomplish that task, and what subset do I not have already. I think part of the problem with deps is that it's not really designed to answer that context specific question. Sure it can be, if you sit there and reason through the tree and pick the right command line switches, and consult a graph theory text book, and meditate for a while, and then intersect that with your Cellar being careful to track what's already satisifed, etc. etc., but it's a silly enterprise if brew install already knows the exact answer to the context specific question without any other fuss.

**Comment by ilovezfs** [comment]
@xu-cheng :shipit: ?

**Comment by xu-cheng** [comment]
👍

**Comment by UniqMartin** [comment]
Not super happy with the commit subject, but I guess that ship has sailed. I think we prefer something more along the lines of `deps: respect --skip-recommended`.

**Comment by ilovezfs** [comment]
It was specifically `build.with?` deps that were overriding everything else in case that wasn't clear that that's the meaning of "build" in the commit subject.

**Comment by UniqMartin** [comment]
Yes, I realize that my example was poorly chosen. I was more referring to the style and not to the exact contents of my example. Anyway, it has been committed … so that's idle talk.
